started state machine

AlenaNguyen opened the pull request:
**8/9 Status**
- Add Intake Floor Implementation
- Add Baseline State Machine logic (RobotContainer)
- Add Baseline Eject Implementation (May need to change based on how the amp/elevator/drainpipe works)
- Updated Drive command to match how we tested it on 8/9 
-  - Includes full (?!) implementation for DT snapping based on state

TODOS: (Should probably be seperate issues)
- [ ] Finish logic to automatically prep shuffle
- [ ] Add logic to automatically prep amp

BrodyKarr committed:
started making a state machine by using enums and started to make the ground intake command🦟™️

BrodyKarr committed:
made the feeder have a neutral output and tried to define digital input for transfer

AlenaNguyen committed:
finished intake command

also created methods for changing/getting the robot state

AlenaNguyen committed:
merge conflicts ✨

alicekuznetsov commented on src/main/java/frc/robot/subsystems/Intake.java:
```diff
@@ -18,7 +18,7 @@ public Intake() {
   }
 
-  public void setRollerOutput(double speed) {
+  public void setIntakeRollerSpeed(double speed) {
```
If we're going to store the intake speeds as measures, we should have the intake methods take measures as their parameters (not doubles)

alicekuznetsov commented on src/main/java/frc/robot/RobotMap.java:
```diff
@@ -59,6 +59,7 @@ public static class mapClimber {
   // MOTORS: 50 -> 59
   public static class mapTransfer {
     public static final int TRANSFER_MOTOR_CAN = 50;
+    // public static final int DigitalInput = 51;
```
I would uncomment this out and reference it when you create the Digital input in Transfer

alicekuznetsov commented on src/main/java/frc/robot/subsystems/Transfer.java:
```diff
@@ -5,25 +5,37 @@
   public Transfer() {
     feederMotor = new TalonFX(mapTransfer.TRANSFER_MOTOR_CAN, "rio");
     feederMotor.getConfigurator().apply(feederConfig);
+    noteSensor = new DigitalInput(1);
```
See my comment on mapTransfer

alicekuznetsov commented on src/main/java/frc/robot/subsystems/Transfer.java:
```diff
@@ -5,25 +5,37 @@
+  }
+
+  public boolean isGamePieceCollected() {
+    return false; // TODO: placeholder for now until we program the beam break sensor
```
Perfect :)

AlenaNguyen committed:
created eject

AlenaNguyen committed:
changed speed methods to require a Measure instead of double

AlenaNguyen committed:
faewjr igwh

alicekuznetsov commented:
Closes #7
Closes #12

alicekuznetsov committed:
Update drivetrain implementation to match testing findings

alicekuznetsov committed:
Complete drivetrain implementation with State Machine :)

alicekuznetsov commented:
Drivetrain implementation with states was tested in sim and it was good stuff 😋
As for the rest- I honestly just went "get as much of the state machine as you can get done today" and Alena just ran with it (shes the 🐐). I'm not sure how we should separate the rest of the parts issue-wise/pr-wise but I think the start is to get this in main

alicekuznetsov committed:
Merge branch 'main' into State-machine

TaylerUva committed:
removed commands

AlenaNguyen committed:
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaahhhhhhhhhhhhhhhhhh

AlenaNguyen committed:
removed null command whoopsie

TaylerUva commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -8,45 +8,92 @@
     return null;
   }
+
+  public static enum RobotState {
+    NONE,
+    INTAKING,
+    STORE_FEEDER,
+    PREP_SHUFFLE,
+    PREP_SPEAKER,
+    PREP_AMP,
+    PREP_NONE,
+    CLIMBING,
+    SHOOTING,
+    EJECTING
+  }
+
+  public static enum TargetState {
+    NONE,
+    PREP_SHUFFLE,
+    PREP_SPEAKER,
+    PREP_AMP,
+    PREP_NONE
+  }
```
These should be moved into the subsystem

TaylerUva commented:
Code looks great so far! Keep up the great work!

TaylerUva commented on src/main/java/frc/robot/subsystems/StateMachine.java:
```diff
@@ -0,0 +1,93 @@
+          case NONE:
+          case PREP_NONE:
+          case PREP_SHUFFLE:
+          case PREP_SPEAKER:
```
We forgot to add `INTAKING` and `STORE_FEEDER` as valid current states

AlenaNguyen commented:
> Code looks great so far! Keep up the great work!

Yay! And also question: how much do we want to get done in this PR before merging because it's getting pretty chonky 🐹 at the moment

TaylerUva commented:
> > Code looks great so far! Keep up the great work!
> 
> Yay! And also question: how much do we want to get done in this PR before merging because it's getting pretty chonky 🐹 at the moment

Good question. Lets address the requested changes and then we can merge this in and finish adding the remaining states in another PR.

TaylerUva commented on src/main/java/frc/robot/commands/States/Eject.java:
Our state commands should probably match our State name one to one. So Eject.java should be Ejecting.java and IntakeFloor.java should be Intaking.java (if we add separate intaking states later we can rename)

TaylerUva commented on src/main/java/frc/robot/commands/States/Eject.java:
```diff
@@ -0,0 +1,54 @@
+  Transfer subTransfer;
+
+  /** Creates a new Eject. */
+  public Eject(Intake subIntake, Transfer subTransfer, StateMachine subStateMachine) {
```
We should probably try to keep the order of subsystem parameters consistent. I recommend subStateMachine first, then alphabetical for the rest. Which looks like what you've already done except for this file:
<img width="753" alt="Screenshot 2024-08-12 at 9 38 56 AM" src="https://github.com/user-attachments/assets/58b60cb0-e1c5-4465-9d44-0fae839e663e">

TaylerUva reviewed (changes requested):
Code looks really good and most of my suggestions are just nitpicking. We can write the remaining states in another PR. Can we create a new issues for each of the states that remain?

AlenaNguyen committed:
addressing review comments

- moved enum states to subsystem
- renamed commands to match states
- made things more organized✨
- also deleted sensor in transfer cuz alice already coded one in main :3

AlenaNguyen committed:
merge conflicts x2 ✨

TaylerUva commented:
We can also test the state transitions in sim using our print outs

AlenaNguyen committed:
changed game piece detection method to Alice's new method

TaylerUva commented on src/main/java/frc/robot/commands/States/NoneState.java:
```diff
@@ -0,0 +1,62 @@
+  public void initialize() {
+    subStateMachine.setRobotState(RobotState.NONE);
+    subElevator.setElevatorPosition(constElevator.BACKWARD_LIMIT);
+    subIntake.setIntakeRollerSpeed(Units.Percent.of(0));
+    subTransfer.setFeederSpeed(Units.Percent.of(0));
+    subShooter.setDesiredVelocities(Units.RotationsPerSecond.of(0), Units.RotationsPerSecond.of(0));
```
```suggestion
    subIntake.setIntakeRollerSpeed(Units.Percent.zero());
    subTransfer.setFeederSpeed(Units.Percent.zero());
    subShooter.setDesiredVelocities(Units.RotationsPerSecond.zero(), Units.RotationsPerSecond.zero());
```
Units have special `.zero()` and `.one()` methods that we can use. `.zero()` will be nice for zeroing thing out since it is very explicit and easy to read. We don't strictly need to do this though

AlenaNguyen committed:
MOREE buttons :3

TaylerUva commented on src/main/java/frc/robot/commands/States/StoreFeeder.java:
```diff
@@ -0,0 +1,51 @@
+  @Override
+  public void initialize() {
+    subStateMachine.setRobotState(RobotState.STORE_FEEDER);
+    subIntake.setIntakeRollerSpeed(Units.Percent.of(0));
+    subTransfer.setFeederSpeed(Units.Percent.of(0));
```
```suggestion
    subIntake.setIntakeRollerSpeed(Units.Percent.zero());
    subTransfer.setFeederSpeed(Units.Percent.zero());
```
Units have special .zero() and .one() methods that we can use. .zero() will be nice for zeroing thing out since it is very explicit and easy to read. We don't strictly need to do this though

TaylerUva committed:
set sim

AlenaNguyen committed:
moved setting targetState to a button press instead of command

AlenaNguyen committed:
merge conflicts (sim edition)

AlenaNguyen commented on src/main/java/frc/robot/subsystems/StateMachine.java:
```diff
@@ -0,0 +1,122 @@
+
+  public Command tryState(RobotState desiredState, StateMachine subStateMachine, Elevator subElevator, Intake subIntake,
+      Transfer subTransfer, Shooter subShooter) {
+    switch (desiredState) {
+      case INTAKING:
+        return new Intaking(subStateMachine, subIntake, subTransfer);
+      // switch (currentState) {
+      // case NONE:
+      // case SHOOTING:
+      // return new Intaking(subStateMachine, subIntake, subTransfer);
+      // }
+
+      case STORE_FEEDER:
+        return new StoreFeeder(subStateMachine, subIntake, subTransfer);
+      // switch (currentState) {
+      // case INTAKING:
+      // return new StoreFeeder(subStateMachine, subIntake, subTransfer);
+      // }
+
+      case EJECTING:
+        return new Ejecting(subStateMachine, subIntake, subTransfer);
+      // switch (currentState) {
+      // case NONE:
+      // case INTAKING:
+      // case STORE_FEEDER:
+      // case PREP_NONE:
+      // case PREP_SHUFFLE:
+      // case PREP_SPEAKER:
+      // return new Ejecting(subStateMachine, subIntake, subTransfer);
+      // }
```
So ze problem that I was stuck on for like 3 hours:
I found out that the checks for a valid _currentState_ in the TryState method isn't working (I think because the whole StateMachine subsystem is being initialized once, so _currentState_ is just stuck at NONE and isn't updating).
Commenting out the _currentState_ checks, the state machine works okay (it can go from NONE to INTAKING to STORE_FEEDER to PREP_SHUFFLE/SPEAKER).
With the _currentState_ checks, the state machine goes from NONE to INTAKING to EJECTING (I think because STORE_FEEDER is not valid since currentState is stuck at NONE, so the logic falls all the way down to EJECTING?).
And ye I'm very stuck 🙃

TaylerUva replied:
Correct assessment of the error. Good job! That is what I determined as well. Adding breaks after the desired state cases will avoid fall through. I have not figured out the solution yet but we could try making things static (static keyword) or try to figure out a way to refresh the try state. Maybe do checks in periodic? Our concept and logic is right but our implantation needs to be rethought.

alicekuznetsov replied:
We had a similar (if not this exact) issue in the 2023_Robot_Code with the stowing command of the arm. Back then, our solution was to always return one giant Commands.either that would have to check every state, but I'm not sure how much I enjoy that implementation. Just my 2 cents from what I've seen us do before
Read it for yourself [here](https://github.com/FRCTeam3255/2023_Robot_Code/blob/main/src/main/java/frc/robot/subsystems/Arm.java) (Line 465)

TaylerUva replied:
> I think because the whole StateMachine subsystem is being initialized once, so _currentState_ is just stuck at NONE and isn't updating

I found the missing ingredient:
```java
gamePieceTrigger
        .onTrue(Commands.deferredProxy(() -> subStateMachine.tryState(RobotState.STORE_FEEDER, subStateMachine,
            subElevator, subIntake, subTransfer,
            subShooter)));
```
The key thing we needed to do was make this a supplier so it was re-evalutated every time instead of once like Alena found.

> I think because STORE_FEEDER is not valid since currentState is stuck at NONE, so the logic falls all the way down to EJECTING?)

Adding breaks after every desiredState case and moving the "default" command outside of the switch statement will fix this.

We can pretty up the code (making it smaller/easier to read) in the next PR but this should make things functional

AlenaNguyen committed:
it does the thing!!

- added lambda and deferredProxy when calling tryState()
- added some more valid states to PREP states (to switch between prep states)

AlenaNguyen committed:
replaced 0's with zero() and moved default command outside switch statement

TaylerUva committed:
Merge branch 'main' into State-machine
